Head manifest endpoint #119

olenagerasimova opened the pull request:
For #119 I've implemented manifest head endpoint, extended test and added todo to add `Docker-Content-Digest` to get and head.

0crat commented:
This pull request [#125](https://github.com/artipie/docker-adapter/issues/125) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/5c642ff9-33be-42ed-a7ea-b0f37707a035); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/9220fdc2-d331-4aeb-a458-32abb413ee3e, version: 0.53.14, hash: ${buildNumber} -->

0crat commented:
@g4s8[/z](https://www.0crat.com/u/g4s8) I see that you unassigned this issue; the order is still assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/dd1ae8d5-732f-4c15-b7cf-745ddbe9f398, version: 0.53.14, hash: ${buildNumber} -->

0crat commented:
This pull request [#125](https://github.com/artipie/docker-adapter/issues/125) is assigned to @olegmoz[/z](https://www.0crat.com/u/olegmoz), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/ac794d06-6861-4401-be70-25d0a84f9050); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; @olegmoz[/z](https://www.0crat.com/u/olegmoz) is not a member of this project yet, but they can request to join, as [§1](http://www.zerocracy.com/policy.html#1) explains; there will be no monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/dce11185-5cc6-4b36-b20f-2f1c46fc1c8b, version: 0.53.14, hash: ${buildNumber} -->

olegmoz commented on src/main/java/com/artipie/docker/http/ManifestEntity.java:
```diff
@@ -78,12 +83,39 @@ private ManifestEntity() {
+            return new AsyncResponse(
+                this.docker.repo(request.name()).manifest(request.reference()).thenCompose(
+                    manifest -> manifest.<CompletionStage<Response>>map(
+                        original -> original.mediaType().thenApply(
+                            type -> new RsWithHeaders(
+                                StandardRs.EMPTY, new ContentType(type)
+                            )
+                        )
```
@olenagerasimova this code duplicates response generation for GET method, I would suggest to reuse `response()` method or introduce `ManifestResponse` class to use in both cases. Adding `Docker-Content-Digest` header will be even easier this way

olenagerasimova replied:
@olegmoz one question - `head` request also can contain `Accept` header, right? So, I should convert manifest to the requested type?

olegmoz replied:
@olenagerasimova good point! Yes, it seems conversion should be performed as well. Docs state that response to HEAD should be the same as GET, but with empty body:
```
[Manifest GET] Fetch the manifest identified by name and reference where reference can be a tag or digest. A HEAD request can also be issued to this endpoint to obtain resource information without receiving all data.
```

olegmoz commented on src/test/java/com/artipie/docker/http/ManifestEntityHeadTest.java:
```diff
@@ -56,34 +62,62 @@ void setUp() {
+     * Assertion.
+     * @param response Actual
+     */
+    private static void assertion(final Response response) {
```
@olenagerasimova I would suggest to change this method name to be a verb, like `assertManifestResponse` maybe. That would describe it better and be more in line with common method naming. You may also consider to solve this code duplication with a `Matcher` instance instead of method, that would be more OOP approach.

olenagerasimova commented:
@olegmoz updated, take a look please. As for code duplication in get/head slices, it seems to me that the only way to avoid it at all is to combine this slices into one, check request method and either add body or not. Otherwise we still have to retrieve `Manifest` instance using async/completable operations chain.

paulodamaso commented on src/test/java/com/artipie/docker/http/ManifestEntityHeadTest.java:
```diff
@@ -56,34 +62,71 @@ void setUp() {
+     * Response matcher.
+     * @since 0.2
+     */
+    private static class ResponseMatcher extends AllOf<Response> {
```
@olenagerasimova I found this `ResponseMatcher` interesting for testing slices. We could  extract it to a class (and maybe put it into `artipie-http` repo) fo reuse, WDYT?

olenagerasimova commented:
@paulodamaso thanks, I like your idea and created https://github.com/artipie/http/issues/153 to introduce such class in http repo. Also I've extracted the class here and added todo to reuse it.

olegmoz commented:
@olenagerasimova thank you, everything looks good

0crat commented:
Job was finished in 23 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

<!-- https://www.0crat.com/footprint/CT2E6TK9B/69c20f22-7455-48d4-a1c3-9ca786180084, version: 0.53.14, hash: ${buildNumber} -->

0crat commented:
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @olegmoz[/z](https://www.0crat.com/u/olegmoz), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/CT2E6TK9B/3e4d4989-423a-4995-b5a1-4ff95faa8077, version: 0.53.14, hash: ${buildNumber} -->

sereshqua commented:
@olegmoz please make sure you will find at least 3 issues during next CR, thanks

sereshqua commented:
@olegmoz ping

sereshqua commented:
@0crat quality acceptable
